compiler/surface/lexer_en.ml

```diff
@@ -500,60 +503,73 @@ let rec lex_code (lexbuf : lexbuf) : token =
-      let pos = lexing_positions lexbuf in
-      if Filename.extension name = ".pdf" then
-        LAW_INCLUDE (Ast.PdfFile ((name, Pos.from_lpos pos), pages))
-      else LAW_INCLUDE (Ast.CatalaFile (name, Pos.from_lpos pos))
-  | Plus '#', Star white_space, Plus (Compl '\n'), Star white_space, '\n' ->
-      L.get_law_heading lexbuf
-  | Plus
-      (* Match non-special characters, i.e. characters that doesn't appear at the start of a
-         previous regexp. *)
-      ( Compl ('#' | '`' | '>')
-      (* Following literals allow to match grave accents as long as they don't conflict with the
-         [BEGIN_CODE] token, i.e. either there are no more than three consecutive ones or they must
-         not be followed by 'catala'. *)
-      | Rep ('`', 1 .. 2), Compl '`'
-      | "```", compl_catala ) ->
-      LAW_TEXT (Utf8.lexeme lexbuf)
+  | '@', Star hspace, "p.", Star hspace, Plus digit ->
+      let s = Utf8.lexeme lexbuf in
+      let i = String.index s '.' in
+      AT_PAGE (int_of_string (String.trim (String.sub s i (String.length s - i))))
+  | Plus (Compl white_space) -> DIRECTIVE_ARG (Utf8.lexeme lexbuf)
+  | Plus hspace -> lex_directive_args lexbuf
+  | '\n' | eof ->
+      L.context := Law;
+      END_DIRECTIVE
+  | _ -> L.raise_lexer_error (Pos.from_lpos prev_pos) prev_lexeme
+
+let rec lex_directive (lexbuf : lexbuf) : token =
+  let prev_lexeme = Utf8.lexeme lexbuf in
+  let prev_pos = lexing_positions lexbuf in
+  match%sedlex lexbuf with
+  | Plus hspace -> lex_directive lexbuf
+  | "Begin", Plus hspace, "metadata" -> BEGIN_METADATA
+  | "End", Plus hspace, "metadata" -> END_METADATA
+  | "Include" -> LAW_INCLUDE
+  | ":" ->
+      L.context := Directive_args;
+      COLON
+  | '\n' | eof ->
+      L.context := Law;
+      END_DIRECTIVE
   | _ -> L.raise_lexer_error (Pos.from_lpos prev_pos) prev_lexeme
 
+(** Main lexing function used outside code blocks *)
+let lex_law (lexbuf : lexbuf) : token =
+  let prev_lexeme = Utf8.lexeme lexbuf in
+  let ((_, start_pos) as prev_pos) = lexing_positions lexbuf in
+  let at_bol = Lexing.(start_pos.pos_bol = start_pos.pos_cnum) in
+  if at_bol then
+    match%sedlex lexbuf with
+    | eof -> EOF
+    | "```catala", Plus white_space ->
+        L.context := Code;
+        Buffer.clear L.code_buffer;
+        BEGIN_CODE
+    | '>' ->
+        L.context := Directive;
+        BEGIN_DIRECTIVE
+    | Plus '#', Star hspace, Plus (Compl '\n'), Star hspace, ('\n' | eof) ->
+        L.get_law_heading lexbuf
+    | _ -> (
+        (* Nested match for lower priority; `_` matches length 0 so we effectively retry the
+           sub-match at the same point *)
+        let lexbuf = lexbuf in
+        (* workaround sedlex bug, see https://github.com/ocaml-community/sedlex/issues/12 *)
+        match%sedlex lexbuf with
+        | Star (Compl '\n'), ('\n' | eof) -> LAW_TEXT (Utf8.lexeme lexbuf)
+        | _ -> L.raise_lexer_error (Pos.from_lpos prev_pos) prev_lexeme)
+  else
```

Comment:
Ok neat, the lexer for law is now aware of beginning of lines. Good ! @EmileRolley we talked about this at some point